libs/wasm/kv/src/lib.rs

```diff
@@ -1,140 +1,92 @@
-use protobuf::{EnumOrUnknown, Message};
+#![allow(non_snake_case)]
+
+use anyhow::anyhow;
+use extism_pdk::*;
+use protobuf::EnumOrUnknown;
 use protos::sp_shared::KVAction;
-use protos::sp_steps_kv::{KVMode, KVStatus, KVStepResponse};
-use protos::sp_wsm::{WASMExitCode, WASMRequest};
+use protos::sp_steps_kv::{KVMode, KVStatus, KVStep, KVStepResponse};
+use protos::sp_wsm::{WASMExitCode, WASMRequest, WASMResponse};
 use streamdal_wasm_detective::detective;
 
-extern "C" {
-    fn kvExists(ptr: *mut u8, length: usize) -> u64;
+#[host_fn]
+extern "ExtismHost" {
+    fn kvExists(step: Protobuf<KVStep>) -> Protobuf<KVStepResponse>;
 }
 
-#[no_mangle]
-pub extern "C" fn f(ptr: *mut u8, length: usize) -> u64 {
-    // Read request
-    let wasm_request = match common::read_request(ptr, length) {
-        Ok(req) => req,
-        Err(e) => {
-            return common::write_error_response(
-                WASMExitCode::WASM_EXIT_CODE_ERROR,
-                format!("unable to parse request: {}", e.to_string()),
-            );
-        }
-    };
-
+#[plugin_fn]
+pub fn f(Protobuf(req): Protobuf<WASMRequest>) -> FnResult<Protobuf<WASMResponse>> {
     // Validate request
-    if let Err(err) = validate_wasm_request(&wasm_request) {
-        return common::write_error_response(
-            WASMExitCode::WASM_EXIT_CODE_ERROR,
-            format!("invalid wasm request: {}", err.to_string()),
-        );
+    if let Err(err) = validate_wasm_request(&req) {
+        return Err(WithReturnCode::new(
+            anyhow!("invalid wasm request: {}", err.to_string()),
+            WASMExitCode::WASM_EXIT_CODE_ERROR as i32,
+        ));
     }
```

Comment:
more nice types in the `f` function signature -- the `Protobuf` extractor/conversion automatically handles the encodings. 

the other changes are mostly converting how errors are returned. 

the `WASMExitCode` will now need to be checked in the SDK side, but shouldn't be a major change.